sw/device/tests/otbn_irq_test.c

```diff
@@ -0,0 +1,167 @@
+#include "sw/device/lib/runtime/otbn.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/entropy_testutils.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTBN_DECLARE_APP_SYMBOLS(err_test);
+
+static const otbn_app_t kAppErrTest = OTBN_APP_T_INIT(err_test);
+
+const test_config_t kTestConfig;
+
+static dif_rv_plic_t plic;
+static volatile bool otbn_finished;
+
+/**
+ * Get OTBN error bits, check this succeeds and code matches `expected_err_bits`
+ */
+static void check_otbn_err_bits(otbn_t *otbn_ctx,
+                                dif_otbn_err_bits_t expected_err_bits) {
+  dif_otbn_err_bits_t otbn_err_bits;
+  CHECK_DIF_OK(dif_otbn_get_err_bits(&otbn_ctx->dif, &otbn_err_bits));
+  CHECK(otbn_err_bits == expected_err_bits,
+        "dif_otbn_get_err_bits() produced unexpected error bits: %x",
+        otbn_err_bits);
+}
+
+/**
+ * Gets the OTBN instruction count, checks that it matches expectations.
+ */
+static void check_otbn_insn_cnt(otbn_t *otbn_ctx, uint32_t expected_insn_cnt) {
+  uint32_t insn_cnt;
+  CHECK_DIF_OK(dif_otbn_get_insn_cnt(&otbn_ctx->dif, &insn_cnt));
+  CHECK(insn_cnt == expected_insn_cnt,
+        "Expected to execute %d instructions, but got %d.", expected_insn_cnt,
+        insn_cnt);
+}
+
+/**
+ * Get OTBN's status and check that it matches expected_status
+ */
+static void check_otbn_status(otbn_t *otbn_ctx,
+                              dif_otbn_status_t expected_status) {
+  dif_otbn_status_t status;
+  CHECK_DIF_OK(dif_otbn_get_status(&otbn_ctx->dif, &status));
+  CHECK(status == expected_status, "Unexpected status: expected %d but got %d.",
+        expected_status, status);
+}
+
+/**
+ * Run a binary on OTBN, waiting for completion by interrupt
+ *
+ * Once the binary has finished, check for expected status, error bits and
+ * instruction count.
+ */
+static void run_test_with_irqs(otbn_t *otbn_ctx, otbn_app_t app,
+                               dif_otbn_status_t expected_status,
+                               dif_otbn_err_bits_t expected_err_bits,
+                               uint32_t expected_insn_cnt) {
+  // Clear the otbn_finished flag: we'll set it in the interrupt handler when
+  // we see the Done interrupt fire.
+  otbn_finished = false;
+
+  CHECK(otbn_load_app(otbn_ctx, app) == kOtbnOk);
+
+  // We don't want to kill everything on a software error, so make sure that SW
+  // errors are not considered fatal.
+  CHECK(dif_otbn_set_ctrl_software_errs_fatal(&otbn_ctx->dif, false) == kDifOk);
+
+  // Enable Done interrupt
+  CHECK_DIF_OK(dif_otbn_irq_set_enabled(&otbn_ctx->dif, kDifOtbnIrqDone,
+                                        kDifToggleEnabled));
+
+  // Start OTBN
+  CHECK(otbn_execute(otbn_ctx) == kOtbnOk);
+
+  // At this point, OTBN should be running. Wait for an interrupt that says
+  // it's done.
+  while (!otbn_finished) {
+    wait_for_interrupt();
+  }
```

Comment:
I wonder if there could be a very slight chance of race condition, when OTBN finishes between `while` check and `wait_for_interrtupt()` call. This would make the test fail on timeout?

Reply:
In case you think that it could be an issue, `IBEX_SPIN_FOR` could be an alternative, the downside of course is figuring out a reasonable timeout value.

Reply:
You're completely right about the race condition: oops! I'll toggle interrupt enables to give myself a critical section.